### ISSUE TITLE
Issues with dds format

### ISSUE THREAD

**Issue by g-adamopoulos** [h3r2tic/tony-mc-mapface#4]
Hello @h3r2tic and thanks for your great work!

Unfortunately I had issues trying to open the .dds LUT in Unity 2022,  therefore I tried converting it, without luck.

* [Compressonator](https://gpuopen.com/compressonator/) crashes when I load the file
* [Nvidia Texture Tools](https://developer.nvidia.com/nvidia-texture-tools-exporter) also fails, but is more verbose about it:
![image](https://user-images.githubusercontent.com/30707361/220437375-93a44f25-6e55-4365-8c59-60c17b0ceeb3.png)

I also tried using the ktx2 file provided in #2, but both software mentioned above had issues opening that file as well.
Even an up-to-date [Unity ktx importer](https://github.com/atteneder/KtxUnity) failed.

Do the above ring a bell ?

Suggestion: A raw data dump of the LUT's contents would be useful, and trivial to directly load into an HLSL Texture3D object. I assume it's a W x H x D array of 32-bit floats ?

Thanks for your time!

**Comment by h3r2tic** [comment]
Ouch! That's painful xD

The DDS is rgb9e5, and maybe that's the problem for those loaders... You could try https://github.com/h3r2tic/tony-mc-mapface/blob/main/OCIO/lut-converter/tony_mc_mapface_f32.dds which is f32 :)

**Comment by Calinou** [comment]
I can confirm Godot 4.0.rc3 can't open the DDS provided on GitHub. Godot doesn't support KTX2 files either, so https://github.com/h3r2tic/tony-mc-mapface/issues/2 isn't an option.

```
Unrecognized or unsupported color layout in DDS 'res://tony_mc_mapface.dds'.
Failed loading resource: res://tony_mc_mapface.dds. Make sure resources have been imported by opening the project in the editor at least once.
editor/editor_node.cpp:1208 - Condition "!res.is_valid()" is true. Returning: ERR_CANT_OPEN
```

> Suggestion: A raw data dump of the LUT's contents would be useful, and trivial to directly load into an HLSL Texture3D object. I assume it's a W x H x D array of 32-bit floats ?

Could a Radiance HDR or OpenEXR image also be created in 2D format (with vertical or horizontal slices)? This would be the easiest to import in Godot as a 3D texture.

**Comment by h3r2tic** [comment]
Does Godot choke on both the rgb9e5 and the rgba32f DDS files?

**Comment by Calinou** [comment]
> Does Godot choke on both the rgb9e5 and the rgba32f DDS files?

Yes, it can't open RGBA32F either. The DDS loader code is here: https://github.com/godotengine/godot/blob/master/modules/dds/texture_loader_dds.cpp

Running from a terminal prints more information:

```
unrecognized fourcc 30315844 format_flags: 4 - rgbbits 0 - red_mask 0 green mask 0 blue mask 0 alpha mask 0
ERROR: Unrecognized or unsupported color layout in DDS 'res://tony_mc_mapface_f32.dds'.
   at: load (modules/dds/texture_loader_dds.cpp:207)

unrecognized fourcc 30315844 format_flags: 4 - rgbbits 0 - red_mask 0 green mask 0 blue mask 0 alpha mask 0
ERROR: Unrecognized or unsupported color layout in DDS 'res://tony_mc_mapface.dds'.
   at: load (modules/dds/texture_loader_dds.cpp:207)
```

GIMP's DDS plugin also can't open both of those formats.

It should be feasible to add support for more DDS formats in a future Godot 4.x release. @fire may be interested in this :slightly_smiling_face:

**Comment by h3r2tic** [comment]
Thanks for confirming! FWIW I exported the DDS using RenderDoc :) Will check whether it can output an unrolled 2D image instead.

**Comment by h3r2tic** [comment]
Perhaps this works?

[hdr_exr.zip](https://github.com/h3r2tic/tony-mc-mapface/files/10809052/hdr_exr.zip)

_No guarantees about the precision of those; I used RenderDoc for the conversion. Please check against some of the images in the README for whether there's no banding, and if the black levels look the same_

**Comment by Calinou** [comment]
> Perhaps this works?

Both import successfully in Godot :slightly_smiling_face: 

However, the green channel's direction is flipped compared to what Godot expects for color correction LUT textures (Y+ is downwards, not upwards).

I've flipped the 48 layers individually in GIMP to make the OpenEXR texture work correctly in Godot. The image is included in this Godot project as `tony_mc_mapface_unrolled_inverted_y.exr`: [color_correction.zip](https://github.com/h3r2tic/tony-mc-mapface/files/10809142/color_correction.zip)

As a result, I'd recommend making Y-flipped textures available as part of the package if possible (that is, with individual layers flipped, **not** the entire image).

**Edit:** The finished project I used for testing is available here: https://github.com/godotengine/godot-demo-projects/pull/857

### Linear tonemapping

![Screenshot_20230223_003050](https://user-images.githubusercontent.com/180032/220787590-90977a8f-694a-4cff-89ed-0cfbbcc4595b.png)

### Linear tonemapping + Tony Mc Mapface as color correction texture

*Clipping appears as white, as linear tonemapping in Godot doesn't support whitepoint adjustments. When used with an appropriate whitepoint value (typically around 6), other tonemapping modes supported by Godot (Reinhard, Filmic, ACES) don't exhibit this white clipping as much when using the color correction texture.*

![Screenshot_20230223_003058](https://user-images.githubusercontent.com/180032/220787592-36cd97b3-959c-4e9a-8aab-f30ad9c40d7b.png)

**Comment by h3r2tic** [comment]
Are you using HDR stimulus to look up into the LUT using the `x/(x+1)` remap from the provided shader? If you're looking it up using clipped LDR data then the output will not be correct.

It might need to be its own separate tone mapper in Godot -- or, if Godot's Reinhard is _exactly_ `x/(x+1)` applied separately to each of the R, G, and B channels, then that might match the input remap the LUT expects.

Also note that the range remap for the LUT is important (the * 47.0/48 + 0.5/48 thing). Perhaps Godot does that for its color correction textures, but perhaps not.

**Comment by Calinou** [comment]
> Are you using HDR stimulus to look up into the LUT using the `x/(x+1)` remap from the provided shader?

No, I'm using Godot's built-in color correction functionality, which doesn't provide a way to do this. It's possible to run [custom post-processing shaders](https://docs.godotengine.org/en/latest/tutorials/shaders/custom_postprocessing.html) but they're slower and have various limitations.

> If you're looking it up using clipped LDR data then the output will not be correct.

Unfortunately, color correction runs *after* tonemapping, and the output data from all tonemappers is clamped between 0 and 1: https://github.com/godotengine/godot/blob/19c9fd6926b958016d562e3b2e93b0e9480300f5/servers/rendering/renderer_rd/shaders/effects/tonemap.glsl#L491-L493

> or, if Godot's Reinhard is exactly x/(x+1) applied separately to each of the R, G, and B channels, then that might match the input remap the LUT expects.

The Reinhard tonemap formula used is here: https://github.com/godotengine/godot/blob/19c9fd6926b958016d562e3b2e93b0e9480300f5/servers/rendering/renderer_rd/shaders/effects/tonemap.glsl#L243-L245

The color correction doesn't look too bad with Reinhard – it helps against the "desaturated" effect that Reinhard often has. It's quite close to Linear without any color correction, but without visible clipping.

### Reinhard tonemapping, whitepoint = 6.0

![Screenshot_20230223_005556](https://user-images.githubusercontent.com/180032/220791445-d672da8e-ef99-4641-85e7-450cbf0a292c.png)

### Reinhard tonemapping, whitepoint = 6.0 + Tony Mc Mapface as color correction texture

![Screenshot_20230223_005604](https://user-images.githubusercontent.com/180032/220791449-5b635d61-7bd4-4539-962a-30f848c71fb6.png)

> Also note that the range remap for the LUT is important (the * 47.0/48 + 0.5/48 thing). Perhaps Godot does that for its color correction textures, but perhaps not.

It's not doing any remap: https://github.com/godotengine/godot/blob/19c9fd6926b958016d562e3b2e93b0e9480300f5/servers/rendering/renderer_rd/shaders/effects/tonemap.glsl#L343-L353

I agree this will need to be a built-in tonemap mode to be implemented correctly in Godot (without the limitations of custom post-processing).

PS: Sorry for going off-topic on this thread, I figure I'd go down this Godot rabbit hole so others don't have to :slightly_smiling_face:

**Comment by h3r2tic** [comment]
Yeah let's not hijack this topic more :sweat_smile: With the formulation that Godot uses for its Reinhard and color correction, TonyMcMapface will not work correctly. It's not really intended to be a color correction LUT; there was a chance it would work if all the planets aligned, but it doesn't seem like they are aligning. Please open a separate issue about Godot if you wish to discuss it further :pray:

**Comment by g-adamopoulos** [comment]
> Ouch! That's painful xD
> 
> The DDS is rgb9e5, and maybe that's the problem for those loaders... You could try https://github.com/h3r2tic/tony-mc-mapface/blob/main/OCIO/lut-converter/tony_mc_mapface_f32.dds which is f32 :)

@h3r2tic Thanks for this!

The NVIDIA Texture Tools app doesn't complain now, but... Unity of course doesn't open it. Unity's .dds importer is quite limited in scope apparently.

That being said, with the knowledge that I'm dealing with 32bit uncompressed RGBA and a proper read of the [microsoft docs](https://learn.microsoft.com/en-us/windows/win32/direct3ddds/dx-graphics-dds-pguide), I wrote a Unity importer, which I will publish today when I clean it up. @Calinou Maybe my C# importer will be a useful starting point for Godot as well.
![image](https://user-images.githubusercontent.com/30707361/220941624-7ed46e3f-7be6-4d8f-a92e-0aad575b8a81.png)

**Comment by g-adamopoulos** [comment]
> Perhaps this works?
> 
> [hdr_exr.zip](https://github.com/h3r2tic/tony-mc-mapface/files/10809052/hdr_exr.zip)
> 
> _No guarantees about the precision of those; I used RenderDoc for the conversion. Please check against some of the images in the README for whether there's no banding, and if the black levels look the same_

I happily confirm that the .hdr and .exr versions **work in Unity** with the right import settings.
![image](https://user-images.githubusercontent.com/30707361/220944742-4bd77dca-ebf3-4260-b5f3-632ee2445e26.png)
![image](https://user-images.githubusercontent.com/30707361/220944871-3c8d3971-31f1-4c3d-bf0b-f8b185ff9a47.png)

**Comment by g-adamopoulos** [comment]
I use this in a Unity shader that blends textures additively over time. All 3 methods (uncompressed .dds imported with my custom importer, .hdr and .exr files) work like a charm!

Before:
![image](https://user-images.githubusercontent.com/30707361/220948204-3c0f2149-d406-40e9-86fe-bd31d00db316.png)
After:
![image](https://user-images.githubusercontent.com/30707361/220948314-cb0948e3-0aa6-43bd-be19-e2ee2cab3d57.png)

**Comment by h3r2tic** [comment]
Nice, happy to hear you've convinced Unity to load it :)

Looks like someone else managed that too, including the rgb9e5 format: https://github.com/Dan3dd/tony-mc-mapface_unity
